File: apps/gharargah/src-tauri/src/host/shell.rs
use serde_json::Value;
use std::process::{Command, Stdio};

use super::uri::file_uri_to_path;

/// Detached launch of an external editor / terminal at a project folder.
pub fn open_in_app(app_id: &str, root_uri: &str) -> Result<(), String> {
    let path = file_uri_to_path(root_uri);
    if path.is_empty() {
        return Err("missing path".into());
    }
    if !std::path::Path::new(&path).exists() {
        return Err(format!("path does not exist: {path}"));
    }

    match app_id {
        "vscode" => {
            try_cmds(&[
                &["code", "-n", &path],
                &["code.cmd", "-n", &path],
            ])
            .or_else(|_| open_macos_app("Visual Studio Code", &path))
        }
        "sublime" => {
            try_cmds(&[&["subl", &path], &["subl.exe", &path]])
                .or_else(|_| open_macos_app("Sublime Text", &path))
        }
        "cursor" => {
            try_cmds(&[&["cursor", &path], &["cursor.cmd", &path]])
                .or_else(|_| open_macos_app("Cursor", &path))
        }
        "ghostty" => {
            let working = format!("--working-directory={path}");
            try_cmds(&[
                &["ghostty", working.as_str()],
                &["open", "-na", "Ghostty", "--args", working.as_str()],
            ])
            .or_else(|_| open_macos_app("Ghostty", &path))
        }
        "kitty" => {
            try_cmds(&[
                &["kitty", "--directory", &path],
                &["kitty", "--single-instance", "--directory", &path],
            ])
            .or_else(|_| open_macos_app("kitty", &path))
        }
        other => Err(format!("unknown app: {other}")),
    }
}

fn try_cmds(attempts: &[&[&str]]) -> Result<(), String> {
    let mut last_err = String::from("no command succeeded");
    for attempt in attempts {
        if attempt.is_empty() {
            continue;
        }
        let (cmd, args) = attempt.split_first().unwrap();
        if *cmd == "open" && !cfg!(target_os = "macos") {
            continue;
        }
        if (*cmd == "code.cmd" || *cmd == "subl.exe" || *cmd == "cursor.cmd") && !cfg!(windows) {
            continue;
        }
        match spawn_detached(cmd, args) {
            Ok(()) => return Ok(()),
            Err(e) => last_err = e,
        }
    }
    Err(last_err)
}

fn open_macos_app(app_name: &str, path: &str) -> Result<(), String> {
    if !cfg!(target_os = "macos") {
        return Err(format!("{app_name} CLI not found"));
    }
    spawn_detached("open", &["-a", app_name, path])
}

fn spawn_detached(program: &str, args: &[&str]) -> Result<(), String> {
    Command::new(program)
        .args(args)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .map(|_| ())
        .map_err(|e| format!("failed to launch {program}: {e}"))
}

pub fn handle(channel: &str, args: &[Value]) -> Result<Value, String> {
    match channel {
        "shell:openInApp" => {
            let app_id = args
                .first()
                .and_then(|v| v.as_str())
                .ok_or("missing appId")?;
            let root_uri = args
                .get(1)
                .and_then(|v| v.as_str())
                .ok_or("missing rootUri")?;
            open_in_app(app_id, root_uri)?;
            Ok(serde_json::json!({ "ok": true }))
        }
        _ => Err(format!("unknown shell channel: {channel}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_unknown_app() {
        let err = open_in_app("nope", "file:///tmp").unwrap_err();
        assert!(err.contains("unknown app"));
    }
}

File: packages/yaade-node-host/src/shell.ts
import { spawn } from "node:child_process"
import fs from "node:fs"
import path from "node:path"
import { uriToPath } from "./paths.js"

function spawnDetached(program: string, args: string[]): void {
  const child = spawn(program, args, {
    detached: true,
    stdio: "ignore",
  })
  child.unref()
}

function tryCmds(attempts: string[][]): void {
  let lastErr = "no command succeeded"
  for (const attempt of attempts) {
    if (attempt.length === 0) continue
    const [cmd, ...args] = attempt
    if (!cmd) continue
    if (cmd === "open" && process.platform !== "darwin") continue
    if (
      (cmd === "code.cmd" || cmd === "subl.exe" || cmd === "cursor.cmd" || cmd === "zed.exe") &&
      process.platform !== "win32"
    ) {
      continue
    }
    try {
      spawnDetached(cmd, args)
      return
    } catch (error) {
      lastErr = String(error)
    }
  }
  throw new Error(lastErr)
}

function openMacosApp(appName: string, path: string): void {
  if (process.platform !== "darwin") throw new Error(`${appName} CLI not found`)
  spawnDetached("open", ["-a", appName, path])
}

export function openInApp(appId: string, rootUri: string): { ok: true } {
  const p = uriToPath(rootUri)
  if (!p) throw new Error("missing path")
  if (!fs.existsSync(p)) throw new Error(`path does not exist: ${p}`)

  switch (appId) {
    case "vscode":
      try {
        tryCmds([
          ["code", "-n", p],
          ["code.cmd", "-n", p],
        ])
      } catch {
        openMacosApp("Visual Studio Code", p)
      }
      break
    case "cursor":
      try {
        tryCmds([
          ["cursor", p],
          ["cursor.cmd", p],
        ])
      } catch {
        openMacosApp("Cursor", p)
      }
      break
    case "emacs":
      try {
        tryCmds([
          ["emacs", p],
          ["emacsclient", "-n", "-a", "", p],
        ])
      } catch {
        openMacosApp("Emacs", p)
      }
      break
    case "sublime":
      try {
        tryCmds([
          ["subl", p],
          ["subl.exe", p],
        ])
      } catch {
        openMacosApp("Sublime Text", p)
      }
      break
    case "zed":
      try {
        tryCmds([
          ["zed", p],
          ["zed.exe", p],
        ])
      } catch {
        openMacosApp("Zed", p)
      }
      break
    case "finder":
      if (process.platform === "darwin") spawnDetached("open", [p])
      else if (process.platform === "win32") spawnDetached("explorer", [p])
      else
        tryCmds([
          ["xdg-open", p],
          ["nautilus", p],
          ["dolphin", p],
        ])
      break
    case "terminal":
      if (process.platform === "darwin") openMacosApp("Terminal", p)
      else if (process.platform === "win32")
        spawnDetached("cmd", ["/c", "start", "cmd", "/k", "cd", "/d", p])
      else
        tryCmds([
          ["x-terminal-emulator", "--working-directory", p],
          ["gnome-terminal", "--working-directory", p],
          ["konsole", "--workdir", p],
        ])
      break
    case "kitty":
      try {
        tryCmds([
          ["kitty", "--directory", p],
          ["kitty", "--single-instance", "--directory", p],
        ])
      } catch {
        openMacosApp("kitty", p)
      }
      break
    case "ghostty": {
      const working = `--working-directory=${p}`
      try {
        tryCmds([
          ["ghostty", working],
          ["open", "-na", "Ghostty", "--args", working],
        ])
      } catch {
        openMacosApp("Ghostty", p)
      }
      break
    }
    case "xcode":
      try {
        tryCmds([["xed", p]])
      } catch {
        openMacosApp("Xcode", p)
      }
      break
    case "intellij":
      try {
        tryCmds([
          ["idea", p],
          ["idea64", p],
          ["intellij-idea-ultimate", p],
          ["intellij-idea-community", p],
        ])
      } catch {
        try {
          openMacosApp("IntelliJ IDEA", p)
        } catch {
          openMacosApp("IntelliJ IDEA CE", p)
        }
      }
      break
    default:
      throw new Error(`unknown app: ${appId}`)
  }
  return { ok: true }
}

/** Reveal a file or folder in the OS file manager (Finder / Explorer / xdg). */
export function revealInFolder(rootUri: string): { ok: true } {
  const p = uriToPath(rootUri)
  if (!p) throw new Error("missing path")
  if (!fs.existsSync(p)) throw new Error(`path does not exist: ${p}`)
  if (process.platform === "darwin") {
    spawnDetached("open", ["-R", p])
  } else if (process.platform === "win32") {
    spawnDetached("explorer", ["/select,", p])
  } else {
    const dir = fs.statSync(p).isDirectory() ? p : path.dirname(p)
    tryCmds([
      ["xdg-open", dir],
      ["nautilus", dir],
      ["dolphin", dir],
    ])
  }
  return { ok: true }
}
